### ISSUE TITLE
Getting DSP info/data leads to unaligned tcache chunk crash on Linux

### ISSUE THREAD

**Issue by GitGhillie** [lebedec/libfmod#20]
examples::core::test_dsp_custom failing on 2.222.5 (commit 1a5de5b) with:
```
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.00s
     Running tests/fmod.rs (target/debug/deps/fmod-df82dc2afa33f964)
tcache_thread_shutdown(): unaligned tcache chunk detected
error: test failed, to rerun pass `-p libfmod --test fmod`
```
In the commit before (4f9383c) this test is passing without issues.

I have the feeling it has something to do with the new `string_buffer!()` and `to_string!()` combo but I'm not sure where to go from here. Issue only occurs on Linux (Ubuntu), not on Windows.

**Comment by lebedec** [comment]
So, we can use Rust for safe and effective development but we can't hide from C legacy (pointers and memory handling) 😄 

Error occurs when memory that has been allocated `does not have properly aligned sizes` 🤔

There are two candidates:
- string_buffer! creates null pointer for empty string, which can be unaligned?
- FMOD_DSP_DESCRIPTION has `paramdesc` wich I recognize as mut pointer to pointers array, but maybe it array of aligned mutable structs

I'm not sure. And unfortunatelly I don't have a Linux environment for testing.

Could you provide more environment info? 
```base
uname -a
rustc --version
rustup toolchain list
```
Could you help to find error by running this code one your environment?:

```rust
#[test]
fn test_dsp_custom() -> Result<(), Error> {
    let system = System::create()?;
    system.init(32, FMOD_INIT_NORMAL, None)?;

    let sound = system.create_sound("./tests/data/Assets/1.ogg", FMOD_LOOP_NORMAL, None)?;
    system.play_sound(sound, None, false)?;

    let volume_desc = DspParameterDesc {
        type_: DspParameterType::Float,
        name: name16("volume"),
        label: name16("%"),
        description: "linear volume in percent".to_string(),
        union: FMOD_DSP_PARAMETER_DESC_UNION {
            floatdesc: FMOD_DSP_PARAMETER_DESC_FLOAT {
                min: 0.0,
                max: 1.0,
                defaultval: 0.42,
                mapping: Default::default(),
            },
        },
    };

    let other_desc = DspParameterDesc {
        type_: DspParameterType::Float,
        name: name16("other"),
        label: name16("%"),
        description: "linear value in percent".to_string(),
        union: FMOD_DSP_PARAMETER_DESC_UNION {
            floatdesc: FMOD_DSP_PARAMETER_DESC_FLOAT {
                min: 0.0,
                max: 1.0,
                defaultval: 0.42,
                mapping: Default::default(),
            },
        },
    };

    struct MyDspData {
        volume: f32,
        other: f32,
    }

    unsafe extern "C" fn create_callback(dsp_state: *mut FMOD_DSP_STATE) -> FMOD_RESULT {
        let data = Box::new(MyDspData {
            volume: 1.0,
            other: 0.0,
        });
        (*dsp_state).plugindata = Box::into_raw(data) as *mut _;
        FMOD_OK
    }

    unsafe extern "C" fn set_parameter_float_callback(
        dsp_state: *mut FMOD_DSP_STATE,
        index: c_int,
        value: c_float,
    ) -> FMOD_RESULT {
        let data = &mut *((*dsp_state).plugindata as *mut MyDspData);
        match index {
            0 => data.volume = value,
            1 => data.other = value,
            _ => unreachable!(),
        }
        FMOD_OK
    }

    unsafe extern "C" fn get_parameter_float_callback(
        dsp_state: *mut FMOD_DSP_STATE,
        index: c_int,
        value: *mut c_float,
        _valuestr: *mut c_char,
    ) -> FMOD_RESULT {
        let data = (*dsp_state).plugindata as *mut MyDspData;
        let data = match index {
            0 => (*data).volume,
            1 => (*data).other,
            _ => unreachable!(),
        };
        value.write(data);
        FMOD_OK
    }
    let mut params = [
        FMOD_DSP_PARAMETER_DESC::try_from(volume_desc).unwrap(),
        FMOD_DSP_PARAMETER_DESC::try_from(other_desc).unwrap(),
    ];
    let dspdesc = ffi::FMOD_DSP_DESCRIPTION {
        pluginsdkversion: 0,
        name: name32("My first DSP unit"),
        version: 0x00010000,
        numinputbuffers: 1,
        numoutputbuffers: 1,
        create: Some(create_callback),
        release: None,
        reset: None,
        read: None,
        process: None,
        setposition: None,
        // *mut *mut pointers
        numparameters: 2,
        paramdesc: [&mut params[0] as *mut _, &mut params[1] as *mut _].as_mut_ptr() as *mut _,
        //
        setparameterfloat: Some(set_parameter_float_callback),
        setparameterint: None,
        setparameterbool: None,
        setparameterdata: None,
        getparameterfloat: Some(get_parameter_float_callback),
        getparameterint: None,
        getparameterbool: None,
        getparameterdata: None,
        shouldiprocess: None,
        userdata: null_mut(),
        sys_register: None,
        sys_deregister: None,
        sys_mix: None,
    };

    // let mydsp = system.create_dsp(dspdesc)?;
    let mydsp = unsafe {
        let mut dsp = null_mut();
        let result = ffi::FMOD_System_CreateDSP(system.as_mut_ptr(), &dspdesc, &mut dsp);
        assert_eq!(result, ffi::FMOD_OK, "dsp ok");
        Dsp::from(dsp)
    };

    let mastergroup = system.get_master_channel_group()?;
    mastergroup.add_dsp(0, mydsp)?;

    for step in 0..5 {
        match step {
            1 => {
                mydsp.set_bypass(true)?;
            }
            2 => {
                mydsp.set_bypass(false)?;
            }
            3 => {
                mydsp.set_parameter_float(0, 0.25)?;
            }
            4 => {
                mydsp.set_parameter_float(1, 0.75)?;
            }
            _ => {}
        }
        thread::sleep(Duration::from_millis(100))
    }

    // string_buffer! 1
    //
    // let info = mydsp.get_parameter_info(0)?;
    // let volume_value_default = unsafe { info.union.floatdesc.defaultval };
    // let (volume_value, _) = mydsp.get_parameter_float(0, 0)?;
    // assert_eq!(volume_value_default, 0.42, "volume default value");
    // assert_eq!(volume_value, 0.25, "volume value");
    // assert_eq!(info.description, "linear volume in percent", "description");

    // string_buffer! 2
    //
    // let info = mydsp.get_parameter_info(1)?;
    // let (other_value, _) = mydsp.get_parameter_float(1, 0)?;
    // assert_eq!(other_value, 0.75, "other value");
    // assert_eq!(info.description, "linear value in percent", "description");

    // string_buffer! 3
    //
    // let (name, version, _, _, _) = mydsp.get_info()?;
    // assert_eq!(name, "My first DSP unit", "dsp info name");
    // assert_eq!(version, 0x00010000, "dsp info version");

    system.release()
}
```

**Comment by GitGhillie** [comment]
> So, we can use Rust for safe and effective development but we can't hide from C legacy (pointers and memory handling) 😄

Haha yeah the technical debt is real :sob: 

Ok so I tried the test you sent and `string_buffer!` sections 1 and 2 work fine, but when I uncomment section 3 (either independently or together with the other sections) then I get the tcache error again.

`uname -a`:
```
Linux gitghillie-pc 6.8.0-45-generic #45-Ubuntu SMP PREEMPT_DYNAMIC Fri Aug 30 12:02:04 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```

`rustc --version`: `rustc 1.81.0 (eeb90cda1 2024-09-04)`

toolchains:
```
stable-x86_64-unknown-linux-gnu (default)
nightly-x86_64-unknown-linux-gnu
```
Hardware is the same as my windows environment btw, I have a dual-boot setup.

Let me know if I can try anything else, I'll try to learn more about the alignment that you mentioned in the meantime :)

**Comment by GitGhillie** [comment]
The new test_channel_group_get_name and test_get_driver_info tests are also failing with the same crash btw (the crash seems to mess  with the test suite so I have to run tests individually to found out which ones are actually failing).

**Comment by GitGhillie** [comment]
PS, doing this seems to work (on both Windows and Linux)
```
macro_rules! string_buffer {
    ($ len : expr) => {
        if $len == 0 {
            std::ptr::null_mut()
        } else {
            CString::from_vec_unchecked(vec![0; $len as usize]).into_raw()
        }
    };
}
```
But doing this also seems to work so I'm not sure how to test for correctness:
```
macro_rules! string_buffer {
    ($ len : expr) => {
        if $len == 0 {
            std::ptr::null_mut()
        } else {
            CString::from_vec_unchecked(vec![0; $len as usize - 2]).into_raw()
        }
    };
}
```

**Comment by lebedec** [comment]
Thank you for troubleshooting. The problem was definitely found. 

But I don't understand why it works like this. CString just chars vec, nothing else. There should be no difference between CString inner buffer and buffer in current implementation of  `string_buffer!`. 

It's very strange that a CString with a smaller size works `$len - 2`. FMOD receives length of C string, that means trailing zero char also counts. For example, If you provide 5 `namelen`, FMOD returns 4 `name` value. Look like Rust CString should be `$len - 1`. 

I'll spend a little more time to figure out how the CString is implemented and why this happens, then I'll apply fix.

**Comment by lebedec** [comment]
The answer is method `as_mut_ptr` is part of the boxed slice implementation. We should call associated function `Box::into_raw` instead. This functions returning raw pointer properly aligned and non-null. Something like this:

```rust
macro_rules! string_buffer {
    ($ len : expr) => {
        if $len == 0 {
            std::ptr::null_mut()
        } else {
            Box::into_raw(vec![0; $len as usize].into_boxed_slice()) as *mut _
        }
    };
}
```

For me It always confusing that same code of "cross-platform" language  has different behaviour on target platforms. Looks like Linux version deallocate buffer memory, but Window not. Or it some sort of undefined behvaiour of incorrect pointers usage, I don't know. 

Anyway, I'll fix this issue and try to add in project auto tests for Linux platform to prevent in future such issues.

**Comment by lebedec** [comment]
Fixed in `libfmod = "2.222.6"`

@GitGhillie thanks for your help 👍

**Comment by GitGhillie** [comment]
Wow yea platform specific issues are the worst... Glad that you were able to find a nice fix & future-proof as well!
I know I said something similar the last time, but I'm confident it will be smooth sailing from here 😁
